### PR TITLE
#479 Allow Iterable, Sequence and Array to be passed to elementsOf

### PR DESCRIPTION
**Pull request by anesabml**
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: misc/deprecated/domain/builders/atrium-domain-builders-common/src/main/kotlin/ch/tutteli/atrium/domain/builders/utils/VarArgHelper.kt [hunk 1]
```diff
@@ -29,12 +30,51 @@ interface VarArgHelper<out T> {
...
+inline fun <reified T> toVarArg(iterableLike: IterableLike): Pair<T, Array<out T>> {
+    when (iterableLike) {
+        is Iterable<*> -> {
+            return iterableToPair(iterableLike.map { it as T })
+        }
+        is Sequence<*> -> {
+            return iterableToPair(iterableLike.map { it as T }.asIterable())
+        }
+        is Array<*> -> {
+            return iterableToPair(iterableLike.map { it as T }.asIterable())
```
Comment (robstoll) [inline misc/deprecated/domain/builders/atrium-domain-builders-common/src/main/kotlin/ch/tutteli/atrium/domain/builders/utils/VarArgHelper.kt:47]: No need for `asIterable` (also below) as map already transforms the Array into a List which is a subtype of Iterable
```suggestion
            return iterableToPair(iterableLike.map { it as T })
```

File: misc/deprecated/domain/builders/atrium-domain-builders-common/src/main/kotlin/ch/tutteli/atrium/domain/builders/utils/VarArgHelper.kt [hunk 2]
```diff
@@ -29,12 +30,51 @@ interface VarArgHelper<out T> {
...
  * requiring `T, vararg T`
  *
  * @throws IllegalArgumentException in case the iterable is empty.
  */
-inline fun <reified T> toVarArg(iterable: Iterable<T>): Pair<T, Array<out T>> {
+inline fun <reified T> toVarArg(iterableLike: IterableLike): Pair<T, Array<out T>> {
+    when (iterableLike) {
+        is Iterable<*> -> {
+            return iterableToPair(iterableLike.map { it as T })
+        }
```
Comment (robstoll) [inline misc/deprecated/domain/builders/atrium-domain-builders-common/src/main/kotlin/ch/tutteli/atrium/domain/builders/utils/VarArgHelper.kt:42]: Please rewrite to
```suggestion
        is Iterable<*> -> iterableToPair(iterableLike.map { it as T })
```
a bit more compact => move return infront of `when` or even better, turn it into an expression body, i.e. 
```
inline fun <reified T> toVarArg(iterableLike: IterableLike): Pair<T, Array<out T>> =
    when (iterableLike) {
      ...
```

**Comment by anesabml** [comment]
Hi, I have a question regarding Implementing `elementsOf` which accepts an `IterableLike` in `iterableContainsInAnyOrderCreators.kt`, `iterableContainsInAnyOrderOnlyCreators.kt` and `iterableContainsInOrderOnlyCreators.kt`. Implementing this requires a change in the `IterableContains` interface, `CheckOptionExtension` and in the `IterableContainsAssertionsBuilder` which implements `IterableContainsAssertions`, so my question is should I modify those files or create a new ones since they will be removed with 1.0.0?

**Comment by robstoll** [comment]
Let's keep Iterable in those cases for now. I'll release 0.13.0 and have already planned to announce the binary breaking change for 0.14.0

I'll let you know once the things underneath the API is ready and then you can continue with Iterable in a second PR

**Comment by anesabml** [comment]
Ok, thanks for the clarification.

**Comment by robstoll** [comment]
@anesabml I took a close look, I don't see why you need to change IterableContains. For instance apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/iterableContainsInAnyOrderCreators.kt => 
you can duplicate `elementsOf` without interfering with the IterableContains interface.

**Comment by anesabml** [comment]
@robstoll Yes indeed I don't need to change those files, it was a misunderstanding, sorry.

**Comment by anesabml** [comment]
Hi @robstoll I want to add tests for the functions I added, but I am not sure how I should do it, will you please point me in the right direction? Thank you.

**Comment by robstoll** [comment]
Very good idea, it seems I forgot to describe it in the issue. Search for elementsOf in *Spec.kt and duplicate the elementsOf accordingly (please also add the TODO to remove the existing with 1.0.0)

**Comment by anesabml** [comment]
great, thank you very much for the explanation.

**Comment by anesabml** [comment]
Hi @robstoll, would you please check the last commit to see if I've added the tests correctly for `IterableContainsInAnyOrderAtLeast1ElementsOfAssertionsSpec`. thank you

**Comment by robstoll** [comment]
@anesabml Looks good to me. Please remove WIP from the title in case you don't intend to do more and I will merge it.

**Comment by anesabml** [comment]
I am working on adding the same functions to api-infix, should I create a different PR for it?

**Comment by robstoll** [comment]
@anesabml ah no, I totally forgot about the infix API. Go ahead :+1:

File: README.md [hunk 3]
```diff
@@ -1363,10 +1363,8 @@ expect(Paths.get("/root/.ssh/config")).isWritable()
 ```text
 expected that subject: /root/.ssh/config        (sun.nio.fs.UnixPath <1234789>)
 ◆ is: writable
-    » failure at parent path: /root        (sun.nio.fs.UnixPath <1234789>)
-      » access was denied
-      » the owner is root, the group is root
-      » the permissions are u=rwx g= o=
+    » no file system entry exists at this location
```
Comment (robstoll) [inline README.md:1366]: please revert this and other changes in the readme (this is caused by the fact that you are not using the same OS).

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/CharSequenceContainsAtLeastAssertionsSpec.kt [hunk 4]
```diff
@@ -36,20 +37,76 @@ class CharSequenceContainsAtLeastAssertionsSpec : Spek({
...
         }
+
         describe("ignoring case elementsOf") {
             it("passing an empty iterable throws an IllegalArgumentException") {
                 expect {
                     expect("test") contains o ignoring case elementsOf emptyList()
                 }.toThrow<IllegalArgumentException> { it messageContains "Iterable without elements are not allowed" }
             }
         }
+
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/CharSequenceContainsAtLeastAssertionsSpec.kt:48]: TODO remove with 1.0.0 (also above)

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableContainsInAnyOrderOnlyElementsOfAssertionsSpec.kt [hunk 5]
```diff
@@ -15,6 +16,24 @@ class IterableContainsInAnyOrderOnlyElementsOfAssertionSpec : Spek({
             }.toThrow<IllegalArgumentException>()
         }
     }
+
+    iterableLikeTypes.forEach { (description, input) ->
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableContainsInAnyOrderOnlyElementsOfAssertionsSpec.kt:20]: add TODO above

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableContainsInOrderOnlyElementsOfAssertionSpec.kt [hunk 6]
```diff
@@ -25,6 +26,42 @@ class IterableContainsInOrderOnlyElementsOfAssertionSpec : Spek({
             }.toThrow<IllegalArgumentException>()
         }
     }
+
+    iterableLikeTypes.forEach { (description, input) ->
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableContainsInOrderOnlyElementsOfAssertionSpec.kt:30]: add TODO remove above

**Comment by robstoll** [comment]
@anesabml many thanks for your first contribution to Atrium :tada: :+1:
There are more first good issues in case you liked it, I could use your help.

**Comment by anesabml** [comment]
Sure I will check out other issues

**Pull request merged**